device/iot-device-samples/device-reconnection-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/IotHubConnectionStatusChangeCallbackHandler.java

```diff
@@ -0,0 +1,95 @@
+                        throw new IllegalStateException(String.format("### DeviceClient cannot be %s with reason %s", status, statusChangeReason));
+                }
+            case DISCONNECTED_RETRYING:
+                switch (statusChangeReason) {
+                    case NO_NETWORK:
+                    case COMMUNICATION_ERROR:
+                        log.debug("### The DeviceClient is retrying based on the retry policy. Do NOT close or open the DeviceClient instance");
+                        log.debug("### The DeviceClient can still queue messages and report properties, but they won't be sent until the connection is established.");
+                        return;
+                    case CONNECTION_OK:
+                    case BAD_CREDENTIAL:
+                    case EXPIRED_SAS_TOKEN:
+                    case RETRY_EXPIRED:
+                    case CLIENT_CLOSE:
+                    default:
+                        throw new IllegalStateException(String.format("### DeviceClient cannot be %s with reason %s", status, statusChangeReason));
+                }
+            case DISCONNECTED:
+                switch (statusChangeReason) {
+                    case CLIENT_CLOSE:
+                        log.debug("### The DeviceClient has been closed gracefully. You can reopen by calling open() on this client.");
+                        return;
+                    case BAD_CREDENTIAL:
+                    case EXPIRED_SAS_TOKEN:
+                        log.warn("### The supplied credentials were invalid. Fix the input and create a new device client instance.");
+                        return;
+                    case RETRY_EXPIRED:
+                        log.warn("### The DeviceClient has been disconnected because the retry policy expired. Can be reopened by closing and then opening the instance.");
+                        if (Application.reconnectIndefinitely) {
+                            new Thread(new Runnable() {
+
+                                @Override
+                                public void run() {
+                                    clientManager.reconnect();
+                                }
+                            }).start();
+                        }
+                        return;
+                    case COMMUNICATION_ERROR:
+                        log.warn("### The DeviceClient has been disconnected due to a non-retryable exception. Inspect the throwable for details.");
+                        log.warn("### The DeviceClient can be reopened by closing and then opening the instance.");
+                        if (Application.reconnectIndefinitely) {
+                            new Thread(new Runnable() {
+
+                                @Override
+                                public void run() {
+                                    clientManager.reconnect();
+                                }
+                            }).start();
+                        }
```

Comment:
This can be factored out to a function since both COMMUNICATION_ERROR and RETRY_EXPIRED both use it